client-java/src/main/java/uk/gov/gchq/palisade/client/internal/resource/WebSocketListener.java

```diff
@@ -0,0 +1,288 @@
+        Optional<Object> getBody();
+
+    }
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(WebSocketListener.class);
+
+    private final ObjectMapper objectMapper;
+    private final Consumer<WebSocketMessage> handler;
+    private final String token;
+
+    /**
+     * A {@code ResourceClient} manages the passing of messages to/from a websocket
+     * server
+     *
+     * @param setup The setup for this listener
+     */
+    public WebSocketListener(final ResourceClientListenerSetup setup) {
+        Checks.checkNotNull(setup);
+        this.token = setup.getToken();
+        this.handler = setup.getEventsHandler();
+        this.objectMapper = setup.getObjectMapper();
+    }
+
+    /**
+     * Helper method to create a {@link Item} using a builder function
+     *
+     * @param func The builder function
+     * @return a newly created {@code RequestId}
+     */
+    @SuppressWarnings("java:S3242")
+    public static WebSocketListener createResourceClientListener(
+        final UnaryOperator<ResourceClientListenerSetup.Builder> func) {
+        return new WebSocketListener(func.apply(new ResourceClientListenerSetup.Builder()).build());
+    }
+
+    @Override
+    public void onError(final WebSocket ws, final Throwable error) {
+        LOGGER.debug("A {} exception was thrown.", error.getCause().getClass().getSimpleName());
+        LOGGER.debug("Message: {}", error.getLocalizedMessage());
+    }
+
+    @Override
+    public void onOpen(final WebSocket ws) {
+        Listener.super.onOpen(ws);
+        LOGGER.debug("WebSocket Listener has been opened for requests.");
+        send(ws, b -> b.type(MessageType.CTS));
+    }
+
+    @Override
+    public CompletionStage<?> onText(final WebSocket ws, final CharSequence data, final boolean last) {
+
+        var text = data.toString();
+
+        ws.request(1); // omit this and no methods are called listener
+
+        Item message;
+        try {
+            message = objectMapper.readValue(text, Item.class);
+        } catch (JsonProcessingException e) {
+            onError(ws, e);
+            return null;
+        }
+
+        LOGGER.debug("<-- {}", message);
+
+        var type = message.getType();
+
+        if (type == MessageType.RESOURCE) {
+
+            var body = message.getBody().orElseThrow(() -> new WebSocketMessageException(message));
+            var item = objectMapper.convertValue(body, ResourceMessage.class);
+            emit(item);
+
+        } else if (type == MessageType.COMPLETE) {
+
+            var item = WebSocketMessage.createComplete(b -> b.token(token));
+            emit(item);
+
+        } else if (type == MessageType.ERROR) {
+
+            var body = message.getBody().orElseThrow(() -> new WebSocketMessageException(message));
+            var item = objectMapper.convertValue(body, ErrorMessage.class);
+            emit(item);
+
+        } else {
+            LOGGER.warn("Ignoring unsupported {} message type", type);
+        }
```

Comment:
`if..else` block are better here as there are multiple statements per type. It would look a bit messy and too easy to miss a `break` with a `switch`. 

Would be much better with switch expressions as each case returns `item`. I think these came in with Java 13 though: https://docs.oracle.com/en/java/javase/13/language/switch-expressions.html

Reply:
I've converted the code to a switch and removed the duplicate code. Looks a bit cleaner now.